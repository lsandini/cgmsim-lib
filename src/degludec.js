const logger = require('pino')();
module.exports = function (weight, degludecs) {
	logger.info(degludecs);

	// activities be expressed as U/min !!!
	const timeSinceDegludecAct = degludecs.map(entry => {
		const time = entry.time;
		const insulin = entry.insulin;
		const duration = 42;
		const peak = (duration / 3);
		const tp = peak;
		const td = duration;

		const tau = tp * (1 - tp / td) / (1 - 2 * tp / td);
		const a = 2 * tau / td;
		const S = 1 / (1 - a + (1 + a) * Math.exp(-td / tau));

		return {
			...entry,
			time: time,
			insulin: insulin,
			degludecActivity: (insulin * (S / Math.pow(tau, 2)) * time * (1 - time / td) * Math.exp(-time / tau)) / 60
		};
	});
	logger.info('these are the degludec activities:', timeSinceDegludecAct);

	// compute the aggregated activity of last degludecs in 45 hours

	const lastDegludecs = timeSinceDegludecAct.filter(function (e) {
		return e.time <= 45;
	});
	logger.info('these are the last degludecs and activities:', lastDegludecs);

	const resultDegAct = lastDegludecs.reduce(function (tot, arr) {
		return tot + arr.degludecActivity;
	}, 0);

	logger.info(resultDegAct);
	return resultDegAct;
};