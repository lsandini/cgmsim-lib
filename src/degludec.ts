import logger from './utils';
import { TreatmentDelta } from './Types';
import { Activity } from './utils';

//const logger = pino();
export default function (degludecs: Pick<TreatmentDelta, 'insulin' | 'minutesAgo'>[]): number {
	logger.info(degludecs);

	// activities be expressed as U/min !!!
	const timeSinceDegludecAct = degludecs.map(entry => {
		const hoursAgo = entry.minutesAgo / 60;
		const insulin = entry.insulin;
		const duration = 42;
		const peak = (duration / 3);
		const { activity } = Activity(peak, duration, hoursAgo, insulin);

		return {
			hours: hoursAgo,
			degludecActivity: activity,
		};
	});

	logger.info('these are the degludec activities: %o', timeSinceDegludecAct);

	// compute the aggregated activity of last degludecs in 45 hours

	const lastDegludecs = timeSinceDegludecAct.filter((e) => e.hours <= 45);
	logger.info('these are the last degludecs and activities: %o', lastDegludecs);

	const resultDegAct = lastDegludecs.reduce((tot, arr) => tot + arr.degludecActivity, 0);

	logger.info(resultDegAct);
	return resultDegAct;
};