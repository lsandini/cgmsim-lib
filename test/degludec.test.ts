import degludec from '../src/degludec';
describe('test degludec', () => {
	test('ins:30 minutesAgo:300', () => {
		const insulinActive = degludec([{
			insulin: 30,
			minutesAgo: 300
		}])
		expect(insulinActive).toMatchSnapshot();

	})

	test('insulin 5 min ago are less active then 40 min ago', () => {
		const insulin = 20;

		const r5 = degludec([{
			insulin,
			minutesAgo: 5
		}])
		const r40 = degludec([{
			insulin,
			minutesAgo: 40
		}])
		expect(r5).toBeLessThan(r40);
	});
	test('peak has the greatest activity', () => {
		const insulin = 20;
		const degludecPeakHours = 42 / 3
		const rBeforePeak = degludec([{
			insulin,
			minutesAgo: (degludecPeakHours * 60) + 10
		}]);
		const rPeak = degludec([{
			insulin,
			minutesAgo: (degludecPeakHours * 60)
		}])
		const rAfterPeak = degludec([{
			insulin,
			minutesAgo: (degludecPeakHours * 60) - 10
		}])
		expect(rBeforePeak).toBeLessThan(rPeak);
		expect(rAfterPeak).toBeLessThan(rPeak);
	});
})