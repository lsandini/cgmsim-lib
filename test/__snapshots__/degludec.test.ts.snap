// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test degludec ins:30 minutesAgo:300 1`] = `0.012545195130933117`;
